File: 模板/位运算/线性基.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from operator import *
from typing import *
from queue import *
from string import *
from random import *

DIRS = [(-1, 0), (1, 0), (0, -1), (0, 1)]
Max = lambda a, b: b if b > a else a
Min = lambda a, b: b if b < a else a
INF = float('inf')
MOD = int(1e9 + 7)


# 线性基模板
class XorBasis:
    def __init__(self, n: int):
        self.b = [0] * n

    def insert(self, x: int) -> None:
        b = self.b
        while x:
            i = x.bit_length() - 1  # x 的最高位
            if b[i] == 0:  # x 和之前的基是线性无关的
                b[i] = x  # 新增一个基，最高位为 i
                return
            x ^= b[i]  # 保证参与 max_xor 的基的最高位是互不相同的，方便我们贪心
        # 正常循环结束，此时 x=0，说明一开始的 x 可以被已有基表出，不是一个线性无关基

    def max_xor(self) -> int:
        b = self.b
        res = 0
        # 从高到低贪心：越高的位，越必须是 1
        # 由于每个位的基至多一个，所以每个位只需考虑异或一个基，若能变大，则异或之
        for i in range(len(b) - 1, -1, -1):
            if res ^ b[i] > res:
                res ^= b[i]
        return res


class Solution:
    def maximizeXorAndXor(self, nums: List[int]) -> int:
        n = len(nums)
        sz = max(nums).bit_length()

        # 预处理所有子集的 AND 和 XOR（刷表法）
        u = 1 << n
        sub_and = [0] * u
        sub_xor = [0] * u
        sub_and[0] = -1
        for i, x in enumerate(nums):
            high_bit = 1 << i
            for mask in range(high_bit):
                sub_and[high_bit | mask] = sub_and[mask] & x
                sub_xor[high_bit | mask] = sub_xor[mask] ^ x
        sub_and[0] = 0

        def max_xor2(sub: int) -> int:
            b = XorBasis(sz)
            xor = sub_xor[sub]
            for i, x in enumerate(nums):
                if sub >> i & 1:
                    # 只考虑有偶数个 1 的比特位（xor 在这些比特位上是 0）
                    b.insert(x & ~xor)
            return xor + b.max_xor() * 2

        return max(sub_and[i] + max_xor2((u - 1) ^ i) for i in range(u))

File: 模板/位运算/线性基.cpp
#include <bits/stdc++.h>
using namespace std;
vector<vector<int>> DIRS = {{1, 1}, {1, -1}, {-1, -1}, {-1, 1}};
using ll = long long;
constexpr int MOD = int(1e9 + 7);



// 线性基模板
class XorBasis {
    vector<uint32_t> b;

public:
    XorBasis(int n) : b(n) {}

    void insert(uint32_t x) {
        while (x) {
            int i = bit_width(x) - 1; // x 的最高位
            if (b[i] == 0) { // x 和之前的基是线性无关的
                b[i] = x; // 新增一个基，最高位为 i
                return;
            }
            x ^= b[i]; // 保证参与 max_xor 的基的最高位是互不相同的，方便我们贪心
        }
        // 正常循环结束，此时 x=0，说明一开始的 x 可以被已有基表出，不是一个线性无关基
    }

    uint32_t max_xor() {
        uint32_t res = 0;
        // 从高到低贪心：越高的位，越必须是 1
        // 由于每个位的基至多一个，所以每个位只需考虑异或一个基，若能变大，则异或之
        for (int i = b.size() - 1; i >= 0; i--) {
            res = max(res, res ^ b[i]);
        }
        return res;
    }
};

// lc3630
class Solution {
public:
    long long maximizeXorAndXor(vector<int> &nums) {
        int n = nums.size();
        int sz = bit_width((uint32_t) ranges::max(nums));

        // 预处理所有子集的 AND 和 XOR（刷表法）
        int u = 1 << n;
        vector<int> sub_and(u), sub_xor(u);
        sub_and[0] = -1;
        for (int i = 0; i < n; i++) {
            int x = nums[i];
            int high_bit = 1 << i;
            for (int mask = 0; mask < high_bit; mask++) {
                sub_and[high_bit | mask] = sub_and[mask] & x;
                sub_xor[high_bit | mask] = sub_xor[mask] ^ x;
            }
        }
        sub_and[0] = 0;

        auto max_xor2 = [&](int sub) -> long long {
            XorBasis b(sz);
            int xor_ = sub_xor[sub];
            for (int i = 0; i < n; i++) {
                if (sub >> i & 1) {
                    // 只考虑有偶数个 1 的比特位（xor 在这些比特位上是 0）
                    b.insert(nums[i] & ~xor_);
                }
            }
            return xor_ + b.max_xor() * 2LL;
        };

        long long ans = 0;
        for (int i = 0; i < u; i++) {
            ans = max(ans, sub_and[i] + max_xor2((u - 1) ^ i));
        }
        return ans;
    }
};
